nemo_aligner/utils/utils.py

```diff
@@ -335,6 +336,35 @@ def retrieve_model_state_dict_in_cpu(model, megatron_amp_O2=True):
     return cpu_dict
 
 
+@torch.no_grad()
+def copy_model_states_to_cpu(model, cpu_dict=None, megatron_amp_O2=True, sync=True, alias_non_tensor=False):
+    """This function mutates the cpu_dict object to throw the model states into preallocated tensors(if they exist)
+        for non tensors it will do a deepcopy, unless alias_non_tensor is True
+    """
+    if cpu_dict is None:
+        cpu_dict = {}
+
+    for name, item in model.state_dict().items():
+        if isinstance(item, torch.Tensor):
+            if name not in cpu_dict:
+                cpu_dict[name] = torch.empty(
+                    item.size(), dtype=item.dtype, layout=item.layout, device="cpu", pin_memory=True
+                )
+            cpu_dict[name].copy_(item, non_blocking=False)
```

Comment:
good point, i think it's better to pin this to the sync variable.

if sync is True: non_blocking=True because we'd sync at the end anyway
if sync is False: non_blocking=False because we are not syncing at the end, so it's potentially dangerous

2775e816424dbbb5b6c02f3c3838b962babb9475

Reply:
Works for me. But out of curiosity is there any use case for `sync=False`?

Reply:
I didn't test for this but it might be possible to do this do sync=False, and force pytorch to sync the weights when it's accessed again. I don't trust that logic fully, so always kept sync=True